pkg/statedb/reflector/k8s.go

```diff
@@ -0,0 +1,254 @@
+	if cfg.BufferWaitTime == 0 {
+		return DefaultBufferWaitTime
+	}
+	return cfg.BufferWaitTime
+}
+
+type KubernetesParams[Obj any] struct {
+	cell.In
+
+	Config KubernetesConfig[Obj]
+	Jobs   job.Registry
+	Scope  cell.Scope
+	DB     *statedb.DB
+}
+
+// Kubernetes synchronizes statedb table with an external Kubernetes resource.
+// Returns a 'Source' for starting and stopping it. If the source can be started
+// unconditionally, consider using [KubernetesCell] instead.
+func Kubernetes[Obj any](p KubernetesParams[Obj]) Reflector[Obj] {
+	tr := &k8sSource[Obj]{
+		KubernetesConfig: p.Config,
+		db:               p.DB,
+	}
+	g := p.Jobs.NewGroup(p.Scope)
+	g.Add(job.OneShot("run", tr.run))
+	return g
+}
+
+// KubernetesCell returns a cell that constructs the Kubernetes source and
+// adds it to the the application lifecycle.
+// For dependencies see [KubernetesParams].
+func KubernetesCell[Obj any]() cell.Cell {
+	return cell.Group(
+		cell.ProvidePrivate(Kubernetes[Obj]),
+		cell.Invoke(func(s Reflector[Obj], lc hive.Lifecycle) { lc.Append(s) }),
+	)
+}
+
+type k8sSource[Obj any] struct {
+	KubernetesConfig[Obj]
+
+	db *statedb.DB
+}
+
+func (s *k8sSource[Obj]) run(ctx context.Context, health cell.HealthReporter) error {
+	type entry struct {
+		deleted   bool
+		name      string
+		namespace string
+		obj       Obj
+	}
+	type buffer map[string]entry
+	bufferSize := s.getBufferSize()
+	waitTime := s.getWaitTime()
+	table := s.Table
+
+	transform := s.Transform
+	if transform == nil {
+		// No provided transform function, use the identity function instead.
+		transform = TransformFunc[Obj](func(obj any) (Obj, bool) { return obj.(Obj), true })
+	}
+
+	// Construct a stream of K8s objects, buffered into chunks every 100 milliseconds.
+	// This reduces the number of write transactions required and thus the number of times
+	// readers get woken up, which results in much better overall throughput.
+	src := stream.Buffer(
+		k8sEventObservable(s.ListerWatcher),
+		bufferSize,
+		waitTime,
+
+		// Buffer the events into a map, coalescing them by key.
+		func(buf buffer, ev cacheStoreEvent) buffer {
+			if buf == nil {
+				buf = make(buffer, bufferSize)
+			}
+			var entry entry
+			entry.deleted = ev.Type == cacheStoreDelete
+
+			var key string
+			if d, ok := ev.Obj.(cache.DeletedFinalStateUnknown); ok {
+				key = d.Key
+				entry.namespace, entry.name, _ = cache.SplitMetaNamespaceKey(d.Key)
+
+				entry.obj, ok = transform(d.Obj)
+				if !ok {
+					return buf
+				}
+			} else {
+				meta, err := meta.Accessor(ev.Obj)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: meta.Accessor failed: %s", s, err))
+				}
+				entry.name = meta.GetName()
+				if ns := meta.GetNamespace(); ns != "" {
+					key = ns + "/" + meta.GetName()
+					entry.namespace = ns
+				} else {
+					key = meta.GetName()
+				}
+
+				var ok bool
+				entry.obj, ok = transform(ev.Obj)
+				if !ok {
+					return buf
+				}
```

Comment:
Nit: we can move this outside of the `if-else` since it is executed in both branches.